src/wh_server_crypto.c

```diff
@@ -81,6 +139,88 @@ int wh_Server_HandleCryptoRequest(whServerContext* server,
     case WC_ALGO_TYPE_PK:
         switch (packet->pkAnyReq.type)
         {
+#ifndef NO_RSA
+        case WC_PK_TYPE_RSA_KEYGEN:
+            /* init the rsa key */
+            ret = wc_InitRsaKey_ex(server->crypto->rsa, NULL, INVALID_DEVID);
+            /* make the rsa key with the given params */
+            if (ret == 0) {
+                ret = wc_MakeRsaKey(server->crypto->rsa,
+                    packet->pkRsakgReq.size,
+                    packet->pkRsakgReq.e,
+                    server->crypto->rng);
+            }
+            /* cache the generated key, data will be blown away */
+            if (ret == 0) {
+                ret = hsmCacheKeyRsa(server, server->crypto->rsa);
+            }
+            wc_FreeRsaKey(server->crypto->rsa);
+            if (ret > 0) {
+                /* set the assigned id */
+                packet->pkRsakgRes.keyId = ret;
+                *size = WOLFHSM_PACKET_STUB_SIZE +
+                    sizeof(packet->pkRsakgRes);
+                ret = 0;
+            }
+            break;
+        case WC_PK_TYPE_RSA:
+            switch (packet->pkRsaReq.opType)
+            {
+                case RSA_PUBLIC_ENCRYPT:
+                case RSA_PUBLIC_DECRYPT:
+                case RSA_PRIVATE_ENCRYPT:
+                case RSA_PRIVATE_DECRYPT:
+                    /* in and out are after the fixed size fields */
+                    in = (uint8_t*)(&packet->pkRsaReq + 1);
+                    out = (uint8_t*)(&packet->pkRsaRes + 1);
+                    /* init rsa key */
+                    ret = wc_InitRsaKey_ex(server->crypto->rsa, NULL,
+                        INVALID_DEVID);
+                    /* load the key from the keystore */
+                    if (ret == 0) {
+                        ret = hsmLoadKeyRsa(server, server->crypto->rsa,
+                            packet->pkRsaReq.keyId);
+                    }
+                    /* do the rsa operation */
+                    if (ret == 0) {
+                        field = packet->pkRsaReq.outLen;
+                        ret = wc_RsaFunction( in, packet->pkRsaReq.inLen,
+                            out, &field, packet->pkRsaReq.opType,
+                            server->crypto->rsa, server->crypto->rng);
+                    }
+                    /* free the key */
+                    wc_FreeRsaKey(server->crypto->rsa);
+                    if (ret == 0) {
+                        /*set outLen */
+                        packet->pkRsaRes.outLen = field;
+                        *size = WOLFHSM_PACKET_STUB_SIZE +
+                            sizeof(packet->pkRsaRes) + field;
+                    }
+                    break;
+            }
+            break;
+        case WC_PK_TYPE_RSA_GET_SIZE:
+            /* init rsa key */
+            ret = wc_InitRsaKey_ex(server->crypto->rsa, NULL, INVALID_DEVID);
```

Comment:
Not sure INVALID_DEVID is correct here, but server->dev_id isn't correct either.  Based on the keyId, we'll need to figure out if the key exists in RAM, NVM, or is a HW-only key.  No changes here, but we'll need to resolve the HW-only case soon.  We already have a backlog entry for this.